File: src/test/kotlin/ru/serdtsev/homemoney/moneyoper/MoneyOperTest.java
package ru.serdtsev.homemoney.moneyoper;

import org.apache.commons.lang3.SerializationUtils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import ru.serdtsev.homemoney.account.model.AccountType;
import ru.serdtsev.homemoney.account.model.Balance;
import ru.serdtsev.homemoney.balancesheet.model.BalanceSheet;
import ru.serdtsev.homemoney.moneyoper.model.MoneyOper;
import ru.serdtsev.homemoney.moneyoper.model.MoneyOperItem;
import ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus;

import java.math.BigDecimal;
import java.sql.Date;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus.cancelled;
import static ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus.done;
import static ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus.pending;

class MoneyOperTest {
  private BalanceSheet balanceSheet;
  private Balance cash;
  private Balance checkingAccount;

  @BeforeEach
  void setUp() {
    balanceSheet = BalanceSheet.Companion.newInstance();
    cash = new Balance(UUID.randomUUID(), balanceSheet, AccountType.debit, "Cash", Date.valueOf(LocalDate.now()), false,
        BigDecimal.TEN, "RUB");
    checkingAccount = new Balance(UUID.randomUUID(), balanceSheet, AccountType.debit, "Checking account",
            Date.valueOf(LocalDate.now()), false,BigDecimal.valueOf(1000L), "RUB");
  }

  @Test
  void changeBalancesByExpense() {
    MoneyOper oper = createExpenseFromCash(done);

    oper.changeBalances(false);
    assertEquals(BigDecimal.valueOf(9L), cash.getValue());

    oper.changeBalances(true);
    assertEquals(BigDecimal.valueOf(10L), cash.getValue());
  }

  @Test
  void changeBalancesByIncome() {
    MoneyOper oper = createIncomeToCash(done);

    oper.changeBalances(false);
    assertEquals(BigDecimal.valueOf(11L), cash.getValue());

    oper.changeBalances(true);
    assertEquals(BigDecimal.valueOf(10L), cash.getValue());
  }

  @Test
  void changeBalancesByTransfer() {
    MoneyOper oper = createTransferFromCheckingAccountToCash(done);

    oper.changeBalances(false);
    assertEquals(BigDecimal.valueOf(999L), checkingAccount.getValue());
    assertEquals(BigDecimal.valueOf(11L), cash.getValue());

    oper.changeBalances(true);
    assertEquals(BigDecimal.valueOf(1000L), checkingAccount.getValue());
    assertEquals(BigDecimal.valueOf(10L), cash.getValue());
  }

  @Test
  void complete() {
    MoneyOper oper = createTransferFromCheckingAccountToCash(pending);
    oper.complete();
    assertEquals(BigDecimal.valueOf(999L), checkingAccount.getValue());
    assertEquals(BigDecimal.valueOf(11L), cash.getValue());
    assertEquals(done, oper.getStatus());
  }

  @Test
  void cancel() {
    MoneyOper oper = createTransferFromCheckingAccountToCash(pending);
    oper.complete();
    oper.cancel();
    assertEquals(BigDecimal.valueOf(1000L), checkingAccount.getValue());
    assertEquals(BigDecimal.valueOf(10L), cash.getValue());
    assertEquals(cancelled, oper.getStatus());
  }

  @Test
  @Disabled
  void essentialEquals() {
    MoneyOper origOper = createExpenseFromCash(done);

    MoneyOper oper = SerializationUtils.clone(origOper);
    assertTrue(oper.essentialEquals(origOper));

    MoneyOperItem item = oper.getItems().get(0);
    item.setValue(item.getValue().add(BigDecimal.ONE));
    assertFalse(oper.essentialEquals(origOper));

    oper = SerializationUtils.clone(origOper);
    oper.addItem(cash, BigDecimal.TEN, LocalDate.now(), 0, UUID.randomUUID());
    assertFalse(oper.essentialEquals(origOper));

    origOper = createTransferFromCheckingAccountToCash(done);
    oper = SerializationUtils.clone(origOper);
    oper.getItems().remove(0);
    assertFalse(oper.essentialEquals(origOper));
  }

  private MoneyOper createExpenseFromCash(MoneyOperStatus status) {
    MoneyOper oper = new MoneyOper(UUID.randomUUID(), balanceSheet, status, LocalDate.now(), 0,
        new ArrayList<>(), "", null);
    oper.addItem(cash, BigDecimal.ONE.negate(), LocalDate.now(), 0, UUID.randomUUID());
    return oper;
  }

  private MoneyOper createIncomeToCash(MoneyOperStatus status) {
    MoneyOper oper = new MoneyOper(UUID.randomUUID(), balanceSheet, status, LocalDate.now(), 0,
        new ArrayList<>(), "", null);
    oper.addItem(cash, BigDecimal.ONE, LocalDate.now(), 0, UUID.randomUUID());
    return oper;
  }

  private MoneyOper createTransferFromCheckingAccountToCash(MoneyOperStatus status) {
    MoneyOper oper = new MoneyOper(UUID.randomUUID(), balanceSheet, status, LocalDate.now(), 0,
        new ArrayList<>(), "", null);
    BigDecimal amount = BigDecimal.ONE;
    oper.addItem(checkingAccount, amount.negate(), LocalDate.now(), 0, UUID.randomUUID());
    oper.addItem(cash, amount, LocalDate.now(), 0, UUID.randomUUID());
    return oper;
  }

}

File: src/test/kotlin/ru/serdtsev/homemoney/moneyoper/MoneyOperTest.kt
package ru.serdtsev.homemoney.moneyoper

import org.apache.commons.lang3.SerializationUtils
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import ru.serdtsev.homemoney.account.model.AccountType
import ru.serdtsev.homemoney.account.model.Balance
import ru.serdtsev.homemoney.balancesheet.model.BalanceSheet
import ru.serdtsev.homemoney.moneyoper.model.MoneyOper
import ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus
import java.math.BigDecimal
import java.time.LocalDate
import java.util.*

internal class MoneyOperTest {
    private lateinit var balanceSheet: BalanceSheet
    private lateinit var cash: Balance
    private lateinit var checkingAccount: Balance

    @BeforeEach
    fun setUp() {
        balanceSheet = BalanceSheet()
        cash = Balance(balanceSheet, AccountType.debit, "Cash", BigDecimal.TEN)
        checkingAccount = Balance(balanceSheet, AccountType.debit, "Checking account", BigDecimal.valueOf(1000L))
    }

    @Test
    fun changeBalancesByExpense() {
        val oper = createExpenseFromCash(MoneyOperStatus.done)
        oper.changeBalances(false)
        assertEquals(BigDecimal.valueOf(900L, 2), cash.value)
        oper.changeBalances(true)
        assertEquals(BigDecimal.valueOf(1000L, 2), cash.value)
    }

    @Test
    fun changeBalancesByIncome() {
        val oper = createIncomeToCash(MoneyOperStatus.done)
        oper.changeBalances(false)
        assertEquals(BigDecimal.valueOf(1100L, 2), cash.value)
        oper.changeBalances(true)
        assertEquals(BigDecimal.valueOf(1000L, 2), cash.value)
    }

    @Test
    fun changeBalancesByTransfer() {
        val oper = createTransferFromCheckingAccountToCash(MoneyOperStatus.done)
        oper.changeBalances(false)
        assertEquals(BigDecimal.valueOf(99900L, 2), checkingAccount.value)
        assertEquals(BigDecimal.valueOf(1100L, 2), cash.value)
        oper.changeBalances(true)
        assertEquals(BigDecimal.valueOf(100000L, 2), checkingAccount.value)
        assertEquals(BigDecimal.valueOf(1000L, 2), cash.value)
    }

    @Test
    fun complete() {
        val oper = createTransferFromCheckingAccountToCash(MoneyOperStatus.pending)
        oper.complete()
        assertEquals(BigDecimal.valueOf(99900L, 2), checkingAccount.value)
        assertEquals(BigDecimal.valueOf(1100L, 2), cash.value)
        assertEquals(MoneyOperStatus.done, oper.status)
    }

    @Test
    fun cancel() {
        val oper = createTransferFromCheckingAccountToCash(MoneyOperStatus.pending)
        oper.complete()
        oper.cancel()
        assertEquals(BigDecimal.valueOf(100000L, 2), checkingAccount.value)
        assertEquals(BigDecimal.valueOf(1000L, 2), cash.value)
        assertEquals(MoneyOperStatus.cancelled, oper.status)
    }

    @Test
    @Disabled
    fun essentialEquals() {
        var origOper = createExpenseFromCash(MoneyOperStatus.done)
        var oper = SerializationUtils.clone(origOper)
        Assertions.assertTrue(oper.mostlyEquals(origOper))
        val item = oper.items[0]
        item.value = item.value.add(BigDecimal.ONE)
        Assertions.assertFalse(oper.mostlyEquals(origOper))
        oper = SerializationUtils.clone(origOper)
        oper.addItem(cash, BigDecimal.TEN, LocalDate.now(), 0, UUID.randomUUID())
        Assertions.assertFalse(oper.mostlyEquals(origOper))
        origOper = createTransferFromCheckingAccountToCash(MoneyOperStatus.done)
        oper = SerializationUtils.clone(origOper)
        oper.items.removeAt(0)
        Assertions.assertFalse(oper.mostlyEquals(origOper))
    }

    private fun createExpenseFromCash(status: MoneyOperStatus): MoneyOper {
        val oper = MoneyOper(
            UUID.randomUUID(), balanceSheet, mutableListOf(), status, LocalDate.now(), 0,
            ArrayList(), "", null
        )
        oper.addItem(cash, BigDecimal.ONE.negate(), LocalDate.now(), 0, UUID.randomUUID())
        return oper
    }

    private fun createIncomeToCash(status: MoneyOperStatus): MoneyOper {
        val oper = MoneyOper(balanceSheet, status)
        oper.addItem(cash, BigDecimal.ONE, LocalDate.now(), 0, UUID.randomUUID())
        return oper
    }

    private fun createTransferFromCheckingAccountToCash(status: MoneyOperStatus): MoneyOper {
        val oper = MoneyOper(balanceSheet, status)
        val amount = BigDecimal.ONE
        oper.addItem(checkingAccount, amount.negate(), LocalDate.now(), 0, UUID.randomUUID())
        oper.addItem(cash, amount, LocalDate.now(), 1, UUID.randomUUID())
        return oper
    }
}